File: 프로그래머스/3/92344. 파괴되지 않은 건물/파괴되지 않은 건물.py
def solution(board, skill):
    n, m = len(board), len(board[0])
    checked = [[0] * (m + 1) for _ in range(n + 1)]
    
    for t, *p, d in skill:
        d *= -1 if t == 1 else 1
        x1, y1, x2, y2 = p
        checked[x1][y1] += d
        checked[x2 + 1][y2 + 1] += d
        checked[x2 + 1][y1] -= d
        checked[x1][y2 + 1] -= d
    
    for i in range(1, n + 1):
        for j in range(m):
            checked[i][j] += checked[i - 1][j]
    
    for i in range(n):
        for j in range(1, m + 1):
            checked[i][j] += checked[i][j - 1]
    
    return sum(board[i][j] + checked[i][j] > 0 for i in range(n) for j in range(m))

File: 프로그래머스/3/92344. 파괴되지 않은 건물/파괴되지 않은 건물.java
import java.util.Arrays;

class Solution {
    public int solution(int[][] board, int[][] skill) {
        int n = board.length;
        int m = board[0].length;
        int[][] scores = new int[n + 1][m + 1];
        
        for (int[] row : scores) {
            Arrays.fill(row, 0);
        }
        
        for (int[] s : skill) {
            int r1 = s[1];
            int c1 = s[2];
            int r2 = s[3];
            int c2 = s[4];
            int degree = s[5];
            
            if (s[0] == 1) degree *= -1;
            
            scores[r1][c1] += degree;
            scores[r1][c2 + 1] -= degree;
            scores[r2 + 1][c1] -= degree;
            scores[r2 + 1][c2 + 1] += degree;
        }
        
        for (int x = 1; x < n; x++) {
            for (int y = 0; y < m; y++) {
                scores[x][y] += scores[x - 1][y];
            }
        }
        
        for (int x = 0; x < n; x++) {
            for (int y = 1; y < m; y++) {
                scores[x][y] += scores[x][y - 1];
            }
        }
        
        int answer = 0;
        
        for (int x = 0; x < n; x++) {
            for (int y = 0; y < m; y++) {
                if (board[x][y] + scores[x][y] > 0) {
                    answer++;
                }
            }
        }
        
        return answer;
    }
}